networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BlobSidecarsByRangeMessageHandler.java

```diff
@@ -0,0 +1,283 @@
+              } else {
+                return sendBlobSidecars(requestState);
+              }
+            });
+  }
+
+  private void handleProcessingRequestError(
+      final Throwable error, final ResponseCallback<BlobSidecar> callback) {
+    final Throwable rootCause = Throwables.getRootCause(error);
+    if (rootCause instanceof RpcException) {
+      LOG.trace("Rejecting blob sidecars by range request", error);
+      callback.completeWithErrorResponse((RpcException) rootCause);
+    } else {
+      if (rootCause instanceof StreamClosedException
+          || rootCause instanceof ClosedChannelException) {
+        LOG.trace("Stream closed while sending requested blobs sidecars", error);
+      } else {
+        LOG.error("Failed to process blob sidecars request", error);
+      }
+      callback.completeWithUnexpectedError(error);
+    }
+  }
+
+  @VisibleForTesting
+  class RequestState {
+
+    private final AtomicInteger sentBlobSidecars = new AtomicInteger(0);
+    private final ResponseCallback<BlobSidecar> callback;
+    private final Bytes32 headBlockRoot;
+    private final AtomicReference<UInt64> currentSlot;
+    private final AtomicReference<UInt64> currentIndex;
+    private final UInt64 maxSlot;
+
+    private Optional<SignedBeaconBlock> maybeCurrentBlock = Optional.empty();
+
+    RequestState(
+        final ResponseCallback<BlobSidecar> callback,
+        final Bytes32 headBlockRoot,
+        final UInt64 currentSlot,
+        final UInt64 maxSlot) {
+      this.callback = callback;
+      this.headBlockRoot = headBlockRoot;
+      this.currentSlot = new AtomicReference<>(currentSlot);
+      this.currentIndex = new AtomicReference<>(UInt64.ZERO);
+      this.maxSlot = maxSlot;
+    }
+
+    @VisibleForTesting
+    UInt64 getCurrentSlot() {
+      return currentSlot.get();
+    }
+
+    @VisibleForTesting
+    UInt64 getCurrentIndex() {
+      return currentIndex.get();
+    }
+
+    SafeFuture<Void> sendBlobSidecar(final BlobSidecar blobSidecar) {
+      return callback.respond(blobSidecar).thenRun(sentBlobSidecars::incrementAndGet);
+    }
+
+    SafeFuture<Optional<BlobSidecar>> loadNextBlobSidecar() {
+      // currentBlock is used to avoid querying the combinedChainDataClient for the same slot again
+      if (maybeCurrentBlock.isEmpty()) {
+        return combinedChainDataClient
+            .getBlockAtSlotExact(currentSlot.get(), headBlockRoot)
+            .thenCompose(
+                block -> {
+                  maybeCurrentBlock = block;
+                  return retrieveBlobSidecar();
+                });
+      } else {
+        return retrieveBlobSidecar();
+      }
+    }
+
+    boolean isComplete() {
+      return currentSlot.get().isGreaterThan(maxSlot)
+          || maxRequestSize.isLessThanOrEqualTo(sentBlobSidecars.get());
+    }
+
+    void incrementSlotAndIndex() {
+      if (currentIndex.get().equals(maxBlobsPerBlock.minus(UInt64.ONE))) {
```

Comment:
I am not sure about that actually, there could be less blobs for a block, we need to check how many kzg commitments are in the block and then decide what indexes to retrieve. I think it would be better though to retrieve all blob sidecars for block root in one method call in the combinedChainDataClient (sth like getBlobSidecarsByBlockRoot() which will return all blob sidecars we have in the DB for given root. It will avoid this index incrementation. @tbenr @zilm13 do you have a view on this?

Reply:
what I don't like is always passing from the block. For finalized blobs if we stream blobs by slot we only get blobs we need (we can safely assume it).

For non finalized we still can stream like that and in case we have multiple block roots for the same slot (we have forks) than we drop non-canonical ones.
The alternative is always pass from blocks in for non finalized.


If we look at `BeaconBlocksByRangeMessageHandler` there are these logics in place.

Reply:
Yeah, I've noticed it too. My opinion is leave it as is now and left a backlog to make an optimization skipping unexcited blobs

Reply:
@StefanBratanov @mehdi-aouadi 
I've tried to rework it from the other side. Not tested, but should be close to production.
https://github.com/zilm13/teku/commit/208e424ba40c56c9ad9074c1309e6e4e168b5ae5
What do you think?

Reply:
@zilm13 The approach you suggested makes sense. Using a data structure could be better. I suggest keeping that for a separate PR. Would that be OK for you?
@StefanBratanov @tbenr Do we keep like that for now and improve it in a separate PR or do you prefer considering the kzg commitments count to avoid requesting non existing blobs?

Reply:
As discussed I logged two separate issues:
https://github.com/ConsenSys/teku/issues/6880
https://github.com/ConsenSys/teku/issues/6881
I'll keep this approach for now and we could improve it later.